File: node-architecture/node.go
package node

import (
	"crypto/ecdsa"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"math/big"
	"net"
	"sync"

	"github.com/golang/protobuf/proto"
	"github.com/tendermint/tendermint/crypto"
	"github.com/tendermint/tendermint/libs/log"
)

// Node represents a node in the IBC network
type Node struct {
	id         string
	chainKey   *ecdsa.PrivateKey
	listener   net.Listener
	logger     log.Logger
	packetChan chan *Packet
}

// NewNode creates a new Node instance
func NewNode(id string, chainKey *ecdsa.PrivateKey, logger log.Logger) *Node {
	return &Node{
		id:         id,
		chainKey:   chainKey,
		logger:     logger,
		packetChan: make(chan *Packet, 10),
	}
}

// Start starts the node
func (n *Node) Start() error {
	listener, err := net.Listen("tcp", ":8080")
	if err != nil {
		return err
	}
	n.listener = listener
	go n.handleIncomingPackets()
	return nil
}

// handleIncomingPackets handles incoming packets from other nodes
func (n *Node) handleIncomingPackets() {
	for {
		conn, err := n.listener.Accept()
		if err != nil {
			n.logger.Error(err)
			continue
		}
		go n.handlePacket(conn)
	}
}

// handlePacket handles a single incoming packet
func (n *Node) handlePacket(conn net.Conn) {
	packet := &Packet{}
	err := json.NewDecoder(conn).Decode(packet)
	if err != nil {
		n.logger.Error(err)
		return
	}
	n.packetChan <- packet
}

// SendPacket sends a packet to another node
func (n *Node) SendPacket(packet *Packet, destNodeID string) error {
	conn, err := net.Dial("tcp", fmt.Sprintf("%s:8080", destNodeID))
	if err != nil {
		return err
	}
	defer conn.Close()
	return json.NewEncoder(conn).Encode(packet)
}

// GetPacketChan returns the packet channel
func (n *Node) GetPacketChan() <-chan *Packet {
	return n.packetChan
}

File: node-architecture/node.rs
use std::collections::{HashMap, VecDeque};
use std::net::{TcpListener, TcpStream};
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Serialize, Deserialize};
use serde_json::{json, Value};

use crypto::{Hash, PublicKey, SecretKey};
use crypto::hashes::{sha256, Sha256};

use tokio::prelude::*;
use tokio::sync::mpsc;

// Node struct
pub struct Node {
    id: String,
    chain_key: Arc<SecretKey>,
    listener: TcpListener,
    logger: slog::Logger,
    packet_chan: Arc<Mutex<VecDeque<Packet>>>,
}

impl Node {
    // Create a new Node instance
    pub fn new(id: String, chain_key: Arc<SecretKey>, logger: slog::Logger) -> Self {
        Node {
            id,
            chain_key,
            listener: TcpListener::bind("0.0.0.0:8080").unwrap(),
            logger,
            packet_chan: Arc::new(Mutex::new(VecDeque::new())),
        }
    }

    // Start the node
    pub async fn start(&self) -> Result<(), String> {
        self.listener.listen(10).unwrap();
        let mut incoming = self.listener.incoming();
        while let Some(stream) = incoming.next().await {
            let stream = stream?;
            self.handle_incoming_packet(stream).await?;
        }
        Ok(())
    }

    // Handle an incoming packet
    async fn handle_incoming_packet(&self, stream: TcpStream) -> Result<(), String> {
        let packet: Packet = serde_json::from_reader(stream).unwrap();
        self.packet_chan.lock().unwrap().push_back(packet);
        Ok(())
    }

    // Send a packet to another node
    pub async fn send_packet(&self, packet: Packet, dest_node_id: String) -> Result<(), String> {
        let mut stream = TcpStream::connect(format!("{}:8080", dest_node_id)).await?;
        serde_json::to_writer(&mut stream, &packet)?;
        Ok(())
    }

    // Get the packet channel
    pub fn get_packet_chan(&self) -> Arc<Mutex<VecDeque<Packet>>> {
        self.packet_chan.clone()
    }
}
